program-analysis/echidna/Exercise-X.md

```diff
@@ -0,0 +1,440 @@
+        // pre-conditions
+        uint256 playerBalanceBefore = naughtCoin.balanceOf(player);
+        uint256 bobBalanceBefore = naughtCoin.balanceOf(bob);
+
+        bool success1 = naughtCoin.transferFrom(player, bob, amount);
+        require(success1, "transferFrom failed");
+
+        // post-conditions
+        assert(
+            naughtCoin.balanceOf(player) == playerBalanceBefore - amount &&
+                naughtCoin.balanceOf(bob) == bobBalanceBefore + amount
+        );
+    }
+   ```
+
+ If you run Echidna now, the `playerBalance == initialSupply` property fails.
+ Echidna was able to find a call sequence to invalidate this property.
+
+ ```
+  assertion in sender_balance_is_equal_to_initial_supply(): FAILED! with ErrorRevert                                   
+  │                                                                                                                      │
+  │ Call sequence:                                                                                                       │
+  │ 1.increaseAllowance(0xa329c0648769a73afac7f9381e08fb43dbea72,11579208923731619542357098500868790785326998466564056403│
+  │   9457584007913129639935) from: 0x0000000000000000000000000000000000030000 Time delay: 531977 seconds Block delay:   │
+  │   12066                                                                                                              │
+  │ 2.no_free_tokens_in_transfer_from(9) from: 0x0000000000000000000000000000000000010000 Time delay: 490448 seconds     │
+  │   Block delay: 3753                                                                                                  │
+  │ 3.sender_balance_is_equal_to_initial_supply() from: 0x0000000000000000000000000000000000030000 Time delay: 1 seconds │
+  │   Block delay: 42595                                                                                                 │
+  │ Event sequence:                                                                                                      │
+  │ Panic(1)
+ ```
+
+</details>
+
+#### Fixing a bug
+
+The last step is to fix a bug that you have found. Try to fix it and re-run your properties. Check if they hold.
+
+<details>
+  <summary>Fixing a bug part 1</summary>
+
+```
+function transferFrom(
+        address _from,
+        address _to,
+        uint256 _amount
+    ) public override lockTokens returns (bool) {
+        super.transferFrom(_from, _to, _amount);
+    }
+```
+
+If you run Echidna, you will see that the properties still fail. Why is that?
+
+Add a `Player(address player)` event to your contract and emit it in your properties. Look at the call sequence.
+
+</details>
+
+<details>
+  <summary>Fixing a bug part 2</summary>
+  
+The player's address is `0x0000...3000`. Echidna is making calls from multiple accounts. It was able to increase the allowance of an address `0x000...1000` and make a call to `transferFrom(player, 0x000...1000, amount)`. This is expected!
```

Comment:
```suggestion
The player's address is `0x0000...3000`. Echidna is making calls from multiple accounts. It was able to increase the allowance of an address `0x000...1000` and make a call to `transferFrom(player, 0x000...1000, amount)`. This is expected!

[This happens because of the `multi-abi`](https://github.com/crytic/building-secure-contracts/blob/master/program-analysis/echidna/using-multi-abi.md). 
```